src/components/Input/index.tsx

```diff
@@ -184,43 +185,109 @@ export interface NumberInputProps extends InputProps {
-            target.value.substring(index + 1, target.value.length);
-        }
+    const handleInput = useCallback<React.FormEventHandler<HTMLInputElement>>(
+      (event) => {
+        const target = event.currentTarget;
         if (
-          target.value.includes(".") &&
-          (target.value?.split(".").pop()?.length || 0) > decimals
+          isIMEActive.current ||
+          !isKeyPressing.current ||
+          target.value.split(".").length > 2
         ) {
+          if (lastValue.current) {
+            target.value = lastValue.current;
+          }
+        }
+        const sanitizedValue = sanitizeNumberInput(target.value, decimals);
+        if (target.value !== sanitizedValue) {
+          event.preventDefault();
+          target.value = sanitizedValue;
+        }
+      },
+      [decimals],
+    );
+
+    const handleKeyDown = useCallback<
+      React.KeyboardEventHandler<HTMLInputElement>
+    >(
+      (event) => {
+        const target = event.currentTarget;
+        isKeyPressing.current = true;
+        isIMEActive.current = event.key === "Process";
+        lastValue.current = target.value;
+        if (isIMEActive.current) {
           event.preventDefault();
-          target.value = formatDecimals(target.value, decimals);
         }
-      };
-      elInput?.addEventListener("keydown", handleKeydown);
-      elInput?.addEventListener("keyup", handleKeydown);
-      elInput?.addEventListener("keypress", handleKeydown);
-      return () => {
-        if (elInput) {
-          elInput.removeEventListener("keydown", handleKeydown);
-          elInput.removeEventListener("keyup", handleKeydown);
-          elInput.removeEventListener("keypress", handleKeydown);
+        if (event.key.length === 1 && !event.ctrlKey) {
+          const regex = /[^0-9.]/g;
+          const isAllowedKey = !regex.test(event.key);
+          const selectedText = target.value?.substring(
+            target.selectionStart || 0,
+            target.selectionEnd || 0,
+          );
+          if (
+            !isAllowedKey ||
+            (event.key === "." &&
+              event.currentTarget.value?.includes(".") &&
+              !selectedText.includes("."))
+          ) {
+            event.preventDefault();
+          }
+        }
+        if (onKeyDown) {
+          onKeyDown(event);
         }
-      };
-    }, [decimals]);
+      },
+      [onKeyDown],
+    );
 
-    return <Input ref={inputRef} {...InputProps} />;
+    const handleKeyUp = useCallback<
+      React.KeyboardEventHandler<HTMLInputElement>
+    >(
+      (event) => {
+        isKeyPressing.current = false;
+        if (onKeyUp) {
+          onKeyUp(event);
+        }
+      },
+      [onKeyUp],
+    );
+
+    const handleChange = useCallback<
+      React.ChangeEventHandler<HTMLInputElement>
+    >(
+      (event) => {
+        const target = event.currentTarget;
+        if (target.value) {
+          try {
+            Numeric.parse(target.value);
```

Comment:
isn't it prevented by replacing `Numeric.parse` in `main.tsx`?

Reply:
`override` would be more precise word I guess.

Reply:
thought of naming it to express its general purpose, like executeAndCachError.
did you intend to reduce changes by overriding? or is there something you wanted to prevent?